### PR TITLE
feat(scheduled builds): Add UI for scheduled builds proposal

### PR DESCRIPTION
**Pull request by JayCeeJr**
#### Empty
Screenshot 2023-05-25 at 12 52 19

#### Add Form
Screenshot 2023-05-25 at 12 52 29

#### List
Screenshot 2023-05-25 at 12 53 09

### CODE REVIEW

**Comment by JayCeeJr** [comment]
> is there a visual mock for what this is supposed to look like?

I added some screenshots to the description

File: src/elm/Pages/Schedules/Update.elm [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+                    )
+    in
+    ( { model | schedulesModel = sm }, action )
+
+
+
+-- takes scheduleModel and returns the URL to redirect to
+
+
+deleteScheduleRedirect : Model msg -> String
```
Comment (plyr4) [inline src/elm/Pages/Schedules/Update.elm:269]: is `deleteScheduleRedirect` used? i cant seem to find the reference

File: src/elm/Pages/Schedules/Form.elm [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+{-| eventEnabled : takes event and returns if it is enabled
+-}
+eventEnabled : String -> List String -> Bool
+eventEnabled event =
+    List.member event
+
+
+schedulesDocsURL : String
+schedulesDocsURL =
+    "#TODO"
```
Comment (plyr4) [inline src/elm/Pages/Schedules/Form.elm:231]: could we fill this value with velaDocsUrl?

File: src/elm/Help/Commands.elm [hunk 3]
```diff
@@ -716,6 +715,94 @@ addSecretArgs =
...
+listSchedules org repo =
+    let
+        name =
+            "List Schedules"
+
+        content =
+            Just <| "vela get schedules " ++ repoArgs org repo
+
+        docs =
+            Just "/build/get"
```
Comment (plyr4) [inline src/elm/Help/Commands.elm:734]:
```suggestion
            Just "/schedule/get"
```

File: src/elm/Api.elm [hunk 4]
```diff
@@ -698,3 +708,47 @@ deleteSecret : PartialModel a -> Engine -> Type -> Org -> Key -> Name -> Request
 deleteSecret model engine type_ org key name =
     delete model.velaAPI (Endpoint.Secret engine type_ org key name) Json.Decode.string
         |> withAuth model.session
+
+
+
+-- SCHEDULES
+
+
+{-| getSchedules : fetches vela builds by repository
```
Comment (plyr4) [inline src/elm/Api.elm:717]:
```suggestion
{-| getSchedules : fetches vela schedules by repository
```

File: src/elm/Pages/Schedules/Update.elm [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+            { schedule | name = value }
+
+        "entry" ->
+            { schedule | entry = value }
+
+        _ ->
+            schedule
+
+
+{-| onChangeAllowCommand : updates allow\_command field on schedule update
```
Comment (plyr4) [inline src/elm/Pages/Schedules/Update.elm:112]:
```suggestion
{-| onChangeAllowCommand : updates allow_command field on schedule update
```

File: src/elm/Pages.elm [hunk 6]
```diff
@@ -37,6 +37,8 @@ type Page
     | Build Org Repo BuildNumber FocusFragment
     | BuildServices Org Repo BuildNumber FocusFragment
     | BuildPipeline Org Repo BuildNumber (Maybe ExpandTemplatesQuery) (Maybe Fragment)
+    | Schedule Org Repo ScheduleName
+    | Schedules Org Repo (Maybe Pagination.Page) (Maybe Pagination.PerPage)
```
[the remark this answers is not in the document]
Comment (JayCeeJr) [reply src/elm/Pages.elm:41]: I could go down that route, I tried this since it seemed like a lot less code, and effectively has the same result.

File: src/elm/Pages/Schedules/Form.elm (hunk 2 above)
Comment (JayCeeJr) [reply src/elm/Pages/Schedules/Form.elm]: I put https://go-vela.github.io/docs/usage/schedules/ this follows the same convention as the others

File: src/elm/Pages.elm (hunk 6 above)
[the remark this answers is not in the document]
Comment (plyr4) [reply src/elm/Pages.elm:41]: true, i'd lean towards having more code if it follows the pattern we have closer

**Review by wass3rw3rk — CHANGES_REQUESTED** [review]
feedback inline; looks to be working fine. thanks for the work!

a couple of items:

### docker compose addition
we should add `VELA_SCHEDULE_ALLOWLIST: '*'` to the docker compose for the server here, just as we did in [go-vela/server](https://github.com/go-vela/server/blob/a5fc7c68a372938e5c6ed3be182a7052ea2a9f24/docker-compose.yml#L46)

### design feedback
tl;dr: no edits requested at this time, although it would be neat if some of the following could make its way in prior to the release.

i'm not sure if there are additional plans for the add/edit view for a schedule, but it's a bit bare and geared towards advanced users at the moment. we can probably agree that cron is not the most approachable format for scheduling to start with, so i was really hoping that we can be a bit more helpful here. you will see that a few online editors stuff the raw cron expression under an "Advanced" section while the initial exposure is something more approachable (or there are helpful annotations or links to dedicated cron tools). here is some quick reference material:
- https://crontab.guru/
- https://tool.crontap.com/
- https://crontab.cronhub.io/
- https://github.com/Marak/cron-editor
- https://esarmientoe.github.io/cron-editor/
- https://karoletrych.github.io/vue-cron-editor/
- image
  (ignore the purple banner in the middle)
- image

not saying any of these are perfect but there are some themes/ideas. i think it boils down to these points for the current implementation:
- what is "Entry"? What am I looking at? *
- either a link to an editor or more inline explanation what each field represents, etc *
- quick views to edit common daily, weekly, hourly, etc schedules in more human way
- a cron expression to word translator would be nice
- something that tells you when the current expression would run next (taking into account server time zone) *
- calendar view would be neat (but value-add is probably diminishing there?)

maybe we can add some of the items at minimum and let user feedback guide the rest. i put a `*` next to the items that i view as lowest effort (honoring server timezone could be a little tricky?).

thoughts?

**Comment by JayCeeJr** [comment]
> 

I changed entry to "Cron Expression" and added a help link to https://crontab.guru/
Maybe it would be better to add the cron expression explanation to the vela docs, and put links to the other cron sites there?

**Comment by wass3rw3rk** [comment]
@JayCeeJr thanks for the edits.. agree - it should probably have a place in the docs as well.

**Review by wass3rw3rk — CHANGES_REQUESTED** [review]
missing cypress tests still

File: src/elm/Main.elm [hunk 7]
```diff
@@ -437,6 +338,7 @@ type Msg
     | HooksResponse (Result (Http.Detailed.Error String) ( Http.Metadata, Hooks ))
     | RedeliverHookResponse Org Repo HookNumber (Result (Http.Detailed.Error String) ( Http.Metadata, String ))
     | BuildResponse Org Repo (Result (Http.Detailed.Error String) ( Http.Metadata, Build ))
+    | SchedulesResponse Org Repo (Result (Http.Detailed.Error String) ( Http.Metadata, Schedules ))
```
Comment (plyr4) [inline src/elm/Main.elm:341]: can we move this to below the `-- Schedules` section?

File: src/elm/Main.elm [hunk 8]
```diff
@@ -2015,6 +1932,97 @@ update msg model =
         NoOp ->
             ( model, Cmd.none )
 
+        SchedulesResponse org repo result ->
```
Comment (plyr4) [inline src/elm/Main.elm:1935]: could we move this section of inbound HTTP responses into the `-- Inbound HTTP responses` section? 

they are currently placed in the `-- Other` section

File: src/elm/Main.elm [hunk 9]
```diff
@@ -3463,6 +3517,19 @@ loadRepoSecretsPage model maybePage maybePerPage engine org repo =
     loadRepoSubPage model org repo <| Pages.RepoSecrets engine org repo maybePage maybePerPage
 
 
+{-| loadRepoSchedulesPage : takes model org and repo and loads the page for managing repo secrets
```
Comment (plyr4) [inline src/elm/Main.elm:3520]:
```suggestion
{-| loadRepoSchedulesPage : takes model org and repo and loads the page for managing repo schedules
```

File: src/elm/Main.elm [hunk 10]
```diff
@@ -3618,6 +3685,55 @@ loadAddRepoSecretPage model engine org repo =
     )
 
 
+{-| loadAddRepoSecretPage : takes model engine org and repo and loads the page for adding secrets
```
Comment (plyr4) [inline src/elm/Main.elm:3688]:
```suggestion
{-| loadAddSchedulePage : takes model engine org and repo and loads the page for adding schedules
```

File: src/elm/Main.elm [hunk 11]
```diff
@@ -3618,6 +3685,55 @@ loadAddRepoSecretPage model engine org repo =
...
+                , deleteState = Pages.Schedules.Model.NotAsked_
+            }
+      }
+    , Cmd.batch
+        [ getCurrentUser model
+        ]
+    )
+
+
+{-| loadAddRepoSecretPage : takes model engine org and repo and loads the page for adding secrets
```
Comment (plyr4) [inline src/elm/Main.elm:3712]:
```suggestion
{-| loadEditSchedulePage : takes model engine org and repo and loads the page for adding schedules
```

**Review by plyr4 — CHANGES_REQUESTED** [review]

File: src/elm/Main.elm [hunk 12]
```diff
@@ -2015,6 +1932,97 @@ update msg model =
...
+                    in
+                    ( { model | schedulesModel = updatedSchedulesModel }
+                    , Cmd.none
+                    )
+
+                Err error ->
+                    ( model, addError error )
+
+        AddScheduleResponse response ->
+            --TODO Handle Add
```
Comment (plyr4) [inline src/elm/Main.elm:1969]: could this be removed?

**Review by plyr4 — CHANGES_REQUESTED** [review]

**Review by plyr4 — APPROVED** [review]

**Comment by plyr4** [comment]
thank you for the tweaks, looks good

**Review by wass3rw3rk — APPROVED** [review]
